plugins/formal-verification/formver.plugin/src/org/jetbrains/kotlin/formver/reporting/ErrorReporter.kt

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2010-2023 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.formver.reporting
+
+import org.jetbrains.kotlin.KtSourceElement
+import org.jetbrains.kotlin.contracts.description.EventOccurrencesRange
+import org.jetbrains.kotlin.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.diagnostics.reportOn
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirDiagnosticRenderers
+import org.jetbrains.kotlin.fir.symbols.FirBasedSymbol
+import org.jetbrains.kotlin.fir.types.ConeKotlinType
+import org.jetbrains.kotlin.fir.types.renderReadable
+import org.jetbrains.kotlin.formver.PluginErrors
+import org.jetbrains.kotlin.formver.embeddings.SourceRole
+import org.jetbrains.kotlin.formver.embeddings.SourceRole.ParamFunctionLeakageCheck.fetchLeakingFunction
+import org.jetbrains.kotlin.formver.viper.errors.VerificationError
+
+/**
+ * Abstract class defining a [report] method used to warn users with human-readable messages.
+ */
+abstract class ErrorReporter<S : SourceRole>(
```

Comment:
I don't understand what you mean by

> I defined a new abstract class called `ErrorReporter<S : SourceRole>` that has a `report` function taking a `DiagnosticReporter# as input parameter. The new class is not an interface to reduce `report` function parameters.

These protected members are only used in the children themselves, why not have them be members of the children?